_posts/2021-09-17-the-joy-of-slow-programming.md

```diff
@@ -0,0 +1,160 @@
+3. make each commit perfect
+
+As you may suspect, these three goals are really hard to fulfill all at the same
+time. I'd usually give up at the stage of learning the new technology or setting up
+the test environment. I didn't have any visible results at the beginning so I was
+not motivated to continue with the project. And a similar pattern would repeat
+with many different ideas.
+
+### New idea
+
+This changed when a good friend of mine came to me with an idea for an application
+that he needed. He was searching for a new apartment in Berlin and noticed that
+for state-owned companies (landeseigene Wohnungsbaugesellschaften) offers were
+disappearing very quickly after being published on the website. Often it was a
+matter of 30-60 minutes. My friend would receive an email notification about
+2-3 hours afterwards, but at this point most of the offers were already gone.
+
+His idea was to develop an application that continuously checks the websites of
+these companies and sends a notification as soon as a new offer is available.
+And he needed my help.
+
+### Fast hacking
+
+We made an appointment for a "hacking evening". I was hoping that we can get
+something up and running within a few hours. All in all, the functionality was
+simple web scraping. I wanted to use some existing solution as much as possible.
+I didn't want to reinvent the wheel and I didn't have that much motivation.
+
+After a quick research I picked [Scrapy](https://scrapy.org/) - a scraping
+framework in Python, supported by a commercial platform where the code can
+run. I started putting up some code together as fast as possible. Soon I
+realised that the persistence is an important topic, because we want to
+somehow know which apartment offers are new and which are not, in order to avoid
+sending the same notifications over and over again. I tried to get the code
+running on the provided infrastructure, with each passing minute getting more
+frustrated that it doesn't work yet, and more disconnected from the reality.
+
+While I was getting deeper and deeper into the programming zone, my friend was
+sitting next to me, getting more and more bored, because at some point I
+stopped explaining what I'm currently doing. Not exactly the "hacking evening"
+anyone would wish for. After my frustration reached a certain level and my energy
+dropped to zero I did `rm -rf the-project` and told my friend: "Well, I guess
+that I need to think about it more in the next days".
+
+### New approach - take it slow
+
+The result of this thinking were a few guiding ideas that I decided on:
+1. use ["boring" technologies](/blog/2018/11/22/boring-ruby-code/) that I already
+know - Ruby, Nokogiri, Rails, and Heroku.
+2. no more than one hour of work on the project per day
+3. focus on delivering the smallest possible thing that actually provides some
+value to my friend
+4. strive for simplicity from the technical perspective and not perfection
```

Comment:
It's interesting that you're contrasting "technical simplicity" against "perfection" as if they're opposed: for me technical simplicity only comes at the end of a refactoring/thinking process, and I mentally associate that process with my own perfectionism 😁

Reply:
Hmm... interesting point! 😉 I guess that I meant "first simple thing that works". That doesn't always yield overall architectural simplicity. I'll see if I can re-write the point to make that clearer and still keep it short and crisp 😄